firedrake/ensemble/ensemble_function.py

```diff
@@ -0,0 +1,292 @@
+        for uself, uriesz in zip(self.subfunctions, riesz.subfunctions):
+            uriesz.assign(
+                uself.riesz_representation(
+                    riesz_map=riesz_map, **kwargs))
+        return riesz
+
+    @PETSc.Log.EventDecorator()
+    def assign(self, other, subsets=None):
+        r"""Set the :class:`EnsembleFunction` to the value of another
+        :class:`EnsembleFunction` other.
+
+        Parameters
+        ----------
+
+        other : :class:`EnsembleFunction`
+            The value to assign from.
+
+        subsets : Collection[Optional[:class:`pyop2.types.set.Subset`]]
+            One subset for each local :class:`Function`. None elements
+            will be ignored.  The values of each local function will
+            only be assigned on the nodes on the corresponding subset.
+        """
+        if type(other) is not type(self):
+            raise ValueError(
+                f"Cannot assign {type(self)} from {type(other)}")
+        for i in range(self._fs.nlocal_spaces):
+            self.subfunctions[i].assign(
+                other.subfunctions[i],
+                subset=subsets[i] if subsets else None)
+        return self
+
+    @PETSc.Log.EventDecorator()
+    def copy(self):
+        """
+        Return a deep copy of the :class:`EnsembleFunction`.
+        """
+        new = type(self)(self.function_space())
+        new.assign(self)
+        return new
+
+    @PETSc.Log.EventDecorator()
+    def zero(self, subsets=None):
+        """
+        Set values to zero.
+
+        Parameters
+        ----------
+
+        subsets : Collection[Optional[:class:`pyop2.types.set.Subset`]]
+            One subset for each local :class:`Function`. None elements
+            will be ignored.  The values of each local function will
+            only be zeroed on the nodes on the corresponding subset.
+        """
+        for i in range(self._fs.nlocal_spaces):
+            self.subfunctions[i].zero(
+                subset=subsets[i] if subsets else None)
+        return self
+
+    @PETSc.Log.EventDecorator()
+    def __iadd__(self, other):
+        for us, uo in zip(self.subfunctions, other.subfunctions):
+            us.assign(us + uo)
+        return self
+
+    @PETSc.Log.EventDecorator()
+    def __imul__(self, other):
+        if type(other) is type(self):
+            for us, uo in zip(self.subfunctions, other.subfunctions):
+                us.assign(us*uo)
+        else:
+            for us in self.subfunctions:
+                us *= other
+        return self
+
+    @PETSc.Log.EventDecorator()
+    def __add__(self, other):
+        new = self.copy()
+        new += other
+        return new
+
+    @PETSc.Log.EventDecorator()
+    def __mul__(self, other):
+        new = self.copy()
+        new *= other
+        return new
+
+    @PETSc.Log.EventDecorator()
+    def __rmul__(self, other):
+        return self.__mul__(other)
+
+    @contextmanager
+    def vec(self):
```

Comment:
I *believe* that David thinks we should use `XXX_rw` instead of `XXX` to make things more explicit (in the context of `.data` vs `.data_rw` but the same applies here)

Reply:
I would prefer that too, but see my comment elsewhere about being consistent with Firedrake. If we think the `vec` context manager should be renamed `vec_rw` then I'd prefer it if we changed it in Firedrake _and_ here, rather than just here.